File: 416-PartitionEqualSubsetSum/416-PartitionEqualSubsetSum.java
// Last updated: 10/3/2025, 2:13:23 PM
class Solution {
    
    int sum;
    Boolean[][] checked;
    public boolean canPartition(int[] nums) {
        for(int i: nums){
            sum+=i;
        }
        if(sum%2 != 0 ) return false;
        sum=sum/2; 
        checked=new Boolean[nums.length][sum+1];
        return helper(nums,0,0);
    }

    boolean helper(int[] nums,int idx,int subsetSum){
        if(subsetSum == sum){
            return true;
        }
        if(subsetSum>sum || idx==nums.length){
            return false;
        }
        if(checked[idx][subsetSum]!=null) return checked[idx][subsetSum];
        boolean pick=helper(nums,idx+1,subsetSum+nums[idx]);
        boolean not_pick=helper(nums,idx+1,subsetSum);

        return checked[idx][subsetSum] = pick || not_pick;

    }
}

File: 416-PartitionEqualSubsetSum/416-PartitionEqualSubsetSum.py
# Last updated: 10/3/2025, 2:13:17 PM
class Solution(object):
    dic={}

    def canPartition(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """

        s=0
        for i in range(len(nums)):
            s+=nums[i]
        if(s%2!=0):
            return False
        
        tar=s//2
        self.dic={}
        return self.rec(tar,0,0,nums)

    def rec(self,tar,s,idx,nums):

        if(idx==len(nums)):
            return False

        # print(s+nums[idx])
        if(s+nums[idx]==tar):
            return True
        
        if(s>tar):
            return False

        if((idx,s) in self.dic):
            return self.dic[(idx,s)]
        
        take=self.rec(tar,s+nums[idx],idx+1,nums)
        not_take=self.rec(tar,s,idx+1,nums)
        self.dic[(idx,s)]=take or not_take
        return self.dic[(idx,s)]



